File: infrastructure-dotnet/src/Infrastructure/Stacks/FrontendStack.cs
using System;
using System.Collections.Generic;
using Amazon.CDK;
using Amazon.CDK.AWS.APIGateway;
using Amazon.CDK.AWS.Apigatewayv2;
using Amazon.CDK.AWS.CloudFront;
using Amazon.CDK.AWS.CloudFront.Origins;
using Amazon.CDK.AWS.Cognito;
using Amazon.CDK.AWS.IAM;
using Amazon.CDK.AWS.S3;
using Amazon.CDK.AWS.S3.Deployment;
using Amazon.CDK.AWS.SSM;
using Constructs;

namespace Infrastructure.Stacks
{
    public class FrontendStackProps : StackProps
    {
        public RestApi RestApi { get; set; }
        public WebSocketApi WebsocketApi { get; set; }
        public string CognitoUserPoolId { get; set; }
        public string CognitoDomainPrefix { get; set; }
    }
    public class FrontendStack : Stack
    {
        internal FrontendStack(Construct scope, string id, FrontendStackProps props = null) : base(scope, id, props)
        {
            var cloudfrontOAI = new OriginAccessIdentity(this, "cloudfront-OAI", new OriginAccessIdentityProps()
            {
                Comment = $"OAI for {id}"
            });

            var siteBucket = new Bucket(this, "SiteBucket", new BucketProps()
            {
                PublicReadAccess = false,
                Encryption = BucketEncryption.S3_MANAGED,
                BlockPublicAccess = BlockPublicAccess.BLOCK_ALL,
                RemovalPolicy = RemovalPolicy.DESTROY, // NOT recommended for production use
                AutoDeleteObjects = true // NOT recommended for production use
            });
            siteBucket.AddToResourcePolicy(new PolicyStatement(new PolicyStatementProps()
            {
                Effect = Effect.DENY,
                Principals = new []{new AnyPrincipal()},
                Actions = new []{"s3:*"},
                Resources = new[]{siteBucket.BucketArn},
                Conditions = new Dictionary<string, object>()
                {
                    {"Bool", new Dictionary<string,string> { {"aws:SecureTransport", "false" }}}
                }
            }));
            
            // *** Log bucket for cloudfront access logging
            // *** UNCOMMENT TO ENABLE ACCESS LOGGING
            /*var logBucket = new Bucket(this, "LogBucket", new BucketProps()
            {
                PublicReadAccess = false,
                Encryption = BucketEncryption.S3_MANAGED,
                BlockPublicAccess = BlockPublicAccess.BLOCK_ALL,
                RemovalPolicy = RemovalPolicy.DESTROY, // NOT recommended for production use
                AutoDeleteObjects = true // NOT recommended for production use
            });
            logBucket.AddToResourcePolicy(new PolicyStatement(new PolicyStatementProps()
            {
                Effect = Effect.DENY,
                Principals = new []{new AnyPrincipal()},
                Actions = new []{"s3:*"},
                Resources = new[]{logBucket.BucketArn},
                Conditions = new Dictionary<string, object>()
                {
                    {"Bool", new Dictionary<string,string> { {"aws:SecureTransport", "false" }}}
                }
            }));*/
            
            // Grant access to cloudfront
            siteBucket.AddToResourcePolicy(new PolicyStatement(new PolicyStatementProps()
            {
                Actions = new [] { "s3:GetObject" },
                Resources = new[] { siteBucket.ArnForObjects("*") },
                Principals = new IPrincipal[] {new CanonicalUserPrincipal(cloudfrontOAI.CloudFrontOriginAccessIdentityS3CanonicalUserId)}
            }));
            new CfnOutput(this, "Bucket", new CfnOutputProps() { Value = siteBucket.BucketName });

            var distribution = new Distribution(this, "SiteDistribution", new DistributionProps()
            {
                DefaultBehavior = new BehaviorOptions()
                {
                    Origin = new S3Origin(siteBucket, new S3OriginProps(){ OriginAccessIdentity = cloudfrontOAI}),
                    Compress = true,
                    ViewerProtocolPolicy = ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
                    AllowedMethods = AllowedMethods.ALLOW_ALL
                },
                DefaultRootObject = "index.html",
                MinimumProtocolVersion = SecurityPolicyProtocol.TLS_V1_2_2021,
                ErrorResponses = new IErrorResponse[]
                {
                    new ErrorResponse() { ResponsePagePath = "/error.html", HttpStatus = 404, ResponseHttpStatus = 404 },
                    new ErrorResponse() { HttpStatus = 403, ResponseHttpStatus = 200, ResponsePagePath = "/index.html", Ttl = Duration.Minutes(1)}
                },
                // *** UNCOMMENT TO ENABLE ACCESS LOGGING
                // LogBucket = logBucket,
                // LogFilePrefix = "distribution-access-logs/",
                // LogIncludesCookies = true
            });

            var cachePolicy = new CachePolicy(this, "CachePolicy", new CachePolicyProps()
            {
                HeaderBehavior = CacheHeaderBehavior.AllowList("Authorization"),
                CookieBehavior = CacheCookieBehavior.None(),
                QueryStringBehavior = CacheQueryStringBehavior.None(),
                EnableAcceptEncodingBrotli = true,
                EnableAcceptEncodingGzip = true,
                MinTtl = Duration.Seconds(1),
                MaxTtl = Duration.Seconds(10),
                DefaultTtl = Duration.Seconds(5)
            });
            
            // REST API behaviour matched to "api/*" path
            distribution.AddBehavior("api/*", 
                new HttpOrigin($"{props?.RestApi.RestApiId}.execute-api.{Stack.Of(this).Region}.amazonaws.com", new HttpOriginProps()
                {
                    OriginPath = $"/{props?.RestApi.DeploymentStage.StageName}"
                }), 
                new BehaviorOptions() {  
                    AllowedMethods = AllowedMethods.ALLOW_ALL,
                    CachePolicy = cachePolicy,
                    ViewerProtocolPolicy = ViewerProtocolPolicy.HTTPS_ONLY,
                    Compress = false
                }
            );

            var wsOriginRequestPolicy = new OriginRequestPolicy(this, "webSocketPolicy", new OriginRequestPolicyProps()
            {
                OriginRequestPolicyName = "webSocketPolicy",
                Comment = "A default WebSocket policy",
                CookieBehavior = OriginRequestCookieBehavior.All(),
                HeaderBehavior = OriginRequestHeaderBehavior.AllowList(new[]{ "Sec-WebSocket-Key", "Sec-WebSocket-Version", "Sec-WebSocket-Protocol", "Sec-WebSocket-Accept" }),
                QueryStringBehavior = OriginRequestQueryStringBehavior.None()
            });
            
            // Websocket API behaviour matched to "wss/*" path
            distribution.AddBehavior("wss/*", 
                new HttpOrigin($"{props?.WebsocketApi.ApiId}.execute-api.{Stack.Of(this).Region}.amazonaws.com", new HttpOriginProps()
                {
                    OriginPath = $"/"
                }), 
                new BehaviorOptions() {  
                    AllowedMethods = AllowedMethods.ALLOW_GET_HEAD,
                    CachePolicy = CachePolicy.CACHING_DISABLED,
                    OriginRequestPolicy = wsOriginRequestPolicy,
                    ViewerProtocolPolicy = ViewerProtocolPolicy.HTTPS_ONLY
                }
            );
            
            // Upload the pre-compiled frontend static files
            var bucketDeployment = new BucketDeployment(this, $"DeployApp-{DateTime.Now.ToShortDateString()}", new BucketDeploymentProps()
            {
                Sources = new[]{Source.Asset("../UI/dist/websocket-chat")},
                DestinationBucket = siteBucket,
                Distribution = distribution,
                DistributionPaths = new[]{"/"}
            });
            
            // Retrieving Cognito Userpool from existing resource - resolving circular stack dependency
            var userPool = UserPool.FromUserPoolId(this, "UserPool", props.CognitoUserPoolId);
            var appClient = userPool.AddClient("websocket-frontend", new UserPoolClientOptions()
            {
                OAuth = new OAuthSettings()
                {
                    Flows = new OAuthFlows()
                    {
                        AuthorizationCodeGrant = false,
                        ImplicitCodeGrant = true // return ID/Access tokens in returnURL
                    },
                    Scopes = new []{ OAuthScope.OPENID, OAuthScope.PROFILE, OAuthScope.EMAIL },
                    CallbackUrls = new []{ $"https://{distribution.DistributionDomainName}/callback", "http://localhost:4200/callback"},
                    LogoutUrls = new []{ $"https://${distribution.DistributionDomainName}/login", "http://localhost:4200/callback" }
                },
                IdTokenValidity = Duration.Minutes(720)
            });
            
            // Generate a cognito app client with a returnURL pointing to the Cloudfront distribution url
            var domain = userPool.AddDomain("Domain", new UserPoolDomainOptions()
            {
                CognitoDomain = new CognitoDomainOptions()
                {
                    DomainPrefix = props.CognitoDomainPrefix
                }
            });

            var cognitoSignInUrl = domain.SignInUrl(appClient, new SignInUrlOptions()
            {
                RedirectUri = $"https://{distribution.DistributionDomainName}/callback" // must be a URL configured under 'callbackUrls' with the client
            });

            var signinUrlParameter = new StringParameter(this, "CognitoSigninURLParameter", new StringParameterProps()
            {
                AllowedPattern = ".*",
                Description = "Cognito Singin URL",
                ParameterName = "/prod/cognito/signinurl",
                StringValue = cognitoSignInUrl,
                Tier = ParameterTier.STANDARD,
            });
            
            var websocketUrlParameter = new StringParameter(this, "WebsocketURLParameter", new StringParameterProps()
            {
                AllowedPattern = ".*",
                Description = "Websocket API URL",
                ParameterName = "/prod/websocket/url",
                StringValue = $"wss://{distribution.DistributionDomainName}/wss/",
                Tier = ParameterTier.STANDARD,
            });
            
            var clientIdParameter = new StringParameter(this, "CognitoClientIdParameter", new StringParameterProps()
            {
                AllowedPattern = ".*",
                Description = "Cognito client id",
                ParameterName = "/prod/cognito/clientid",
                StringValue = appClient.UserPoolClientId,
                Tier = ParameterTier.STANDARD,
            });
            
            new CfnOutput(this, "cognitoSigninURL", new CfnOutputProps() { Value = cognitoSignInUrl, Description = "SignIn URL for Cognito Userpool"});
            new CfnOutput(this, "DistributionId", new CfnOutputProps() { Value = distribution.DistributionId});
            new CfnOutput(this, "DistributionURL", new CfnOutputProps() { Value = distribution.DistributionDomainName});
        }
    }
}

File: infrastructure-ts/lib/frontend-stack.ts
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

import { CfnOutput, Duration, RemovalPolicy, Stack, StackProps } from 'aws-cdk-lib'
import { BucketDeployment, Source } from 'aws-cdk-lib/aws-s3-deployment';
import { RestApi } from 'aws-cdk-lib/aws-apigateway';
import { CacheCookieBehavior, CacheHeaderBehavior, CachePolicy, CacheQueryStringBehavior, SecurityPolicyProtocol, ViewerProtocolPolicy } from 'aws-cdk-lib/aws-cloudfront';
import { OAuthScope, UserPool } from 'aws-cdk-lib/aws-cognito';
import { ParameterTier, StringParameter } from 'aws-cdk-lib/aws-ssm';
import { Construct } from 'constructs';
import * as cloudfront from 'aws-cdk-lib/aws-cloudfront';
import * as origins from 'aws-cdk-lib/aws-cloudfront-origins';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as s3 from 'aws-cdk-lib/aws-s3';
import { WebSocketApi } from 'aws-cdk-lib/aws-apigatewayv2';
import { NagSuppressions } from 'cdk-nag';
import { AnyPrincipal, Effect, PolicyStatement } from 'aws-cdk-lib/aws-iam';

export interface FrontendProps extends StackProps {
  restApi: RestApi;
  websocketApi: WebSocketApi;
  cognitoUserPoolId: string;
  cognitoDomainPrefix: string;
}

export class FrontendStack extends Stack {
  constructor(scope: Construct, id: string, props?: FrontendProps) {
    super(scope, id, props);

    const cloudfrontOAI = new cloudfront.OriginAccessIdentity(this, 'cloudfront-OAI', { comment: `OAI for ${id}` });

    // Content bucket
    const siteBucket = new s3.Bucket(this, 'SiteBucket', {
      publicReadAccess: false,
      encryption: s3.BucketEncryption.S3_MANAGED,
      blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
      removalPolicy: RemovalPolicy.DESTROY, // NOT recommended for production use
      autoDeleteObjects: true, // NOT recommended for production use
    });
    siteBucket.addToResourcePolicy(new PolicyStatement({
      effect: Effect.DENY,
      principals: [
        new AnyPrincipal(),
      ],
      actions: [
        "s3:*"
      ],
      resources: [siteBucket.bucketArn],
      conditions: {
        "Bool": { "aws:SecureTransport": "false" },
      },
    }));

    // *** Log bucket for cloudfront access logging
    // *** UNCOMMENT TO ENABLE ACCESS LOGGING
    // const logBucket = new s3.Bucket(this, 'LogBucket', {
    //   publicReadAccess: false,
    //   encryption: s3.BucketEncryption.S3_MANAGED,
    //   blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
    //   removalPolicy: RemovalPolicy.DESTROY, // NOT recommended for production use
    //   autoDeleteObjects: true, // NOT recommended for production use
    // });
    // logBucket.addToResourcePolicy(new PolicyStatement({
    //   effect: Effect.DENY,
    //   principals: [
    //     new AnyPrincipal(),
    //   ],
    //   actions: [
    //     "s3:*"
    //   ],
    //   resources: [logBucket.bucketArn],
    //   conditions: {
    //     "Bool": { "aws:SecureTransport": "false" },
    //   },
    // }));

    // Grant access to cloudfront
    siteBucket.addToResourcePolicy(new PolicyStatement({
      actions: ['s3:GetObject'],
      resources: [siteBucket.arnForObjects('*')],
      principals: [new iam.CanonicalUserPrincipal(cloudfrontOAI.cloudFrontOriginAccessIdentityS3CanonicalUserId)]
    }));
    new CfnOutput(this, 'Bucket', { value: siteBucket.bucketName });

    const distribution = new cloudfront.Distribution(this, 'SiteDistribution', {
      defaultBehavior: { // Default to S3 bucket
        origin: new origins.S3Origin(siteBucket, { originAccessIdentity: cloudfrontOAI }),
        compress: true,
        viewerProtocolPolicy: cloudfront.ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
        allowedMethods: cloudfront.AllowedMethods.ALLOW_ALL,
      },
      defaultRootObject: 'index.html',
      minimumProtocolVersion: SecurityPolicyProtocol.TLS_V1_2_2021,
      errorResponses: [{ responsePagePath: '/error.html', httpStatus: 404, responseHttpStatus: 404 },
      {
        httpStatus: 403,
        responseHttpStatus: 200,
        responsePagePath: '/index.html',
        ttl: Duration.minutes(1),
      }],
      // *** UNCOMMENT TO ENABLE ACCESS LOGGING
      //logBucket: logBucket,
      //logFilePrefix: 'distribution-access-logs/',
      //logIncludesCookies: true
    });
    NagSuppressions.addResourceSuppressions(
      distribution,
      [
        {
          id: 'AwsSolutions-CFR3',
          reason:
            "Access logging is disabled to save cost. It can be re-enabled by uncommenting the code above."
        },
        {
          id: 'AwsSolutions-CFR4',
          reason:
            "TLSv1.1 or TLSv1.2 can be only enforced using a custom certificate with a custom domain alias."
        },
      ],
      true
    );


    // Custom Cloudfront cache policy to forward Authorization header
    const cachePolicy = new CachePolicy(this, 'CachePolicy', {
      headerBehavior: CacheHeaderBehavior.allowList(
        'Authorization',
      ),
      cookieBehavior: CacheCookieBehavior.none(),
      queryStringBehavior: CacheQueryStringBehavior.none(),
      enableAcceptEncodingBrotli: true,
      enableAcceptEncodingGzip: true,
      minTtl: Duration.seconds(1),
      maxTtl: Duration.seconds(10),
      defaultTtl: Duration.seconds(5),
    })

    // REST API behaviour matched to "api/*" path
    distribution.addBehavior('api/*', new origins.HttpOrigin(`${props?.restApi.restApiId}.execute-api.${Stack.of(this).region}.amazonaws.com`, {
      originPath: `/${props?.restApi.deploymentStage.stageName}`
    }), {
      allowedMethods: cloudfront.AllowedMethods.ALLOW_ALL,
      cachePolicy: cachePolicy,
      viewerProtocolPolicy: ViewerProtocolPolicy.HTTPS_ONLY,
      compress: false
    });

    const wsOriginRequestPolicy = new cloudfront.OriginRequestPolicy(this, "webSocketPolicy", {
      originRequestPolicyName: "webSocketPolicy",
      comment: "A default WebSocket policy",
      cookieBehavior: cloudfront.OriginRequestCookieBehavior.all(),
      headerBehavior: cloudfront.OriginRequestHeaderBehavior.allowList("Sec-WebSocket-Key", "Sec-WebSocket-Version", "Sec-WebSocket-Protocol", "Sec-WebSocket-Accept"),
      queryStringBehavior: cloudfront.OriginRequestQueryStringBehavior.none(),
    });

    // Websocket API behaviour matched to "wss/*" path
    distribution.addBehavior('wss/*', new origins.HttpOrigin(`${props?.websocketApi.apiId}.execute-api.${Stack.of(this).region}.amazonaws.com`, {
      originPath: `/`
    }), {
      allowedMethods: cloudfront.AllowedMethods.ALLOW_GET_HEAD,
      cachePolicy: cloudfront.CachePolicy.CACHING_DISABLED,
      originRequestPolicy: wsOriginRequestPolicy,
      viewerProtocolPolicy: ViewerProtocolPolicy.HTTPS_ONLY
    });

    // Upload the pre-compiled frontend static files
    new BucketDeployment(this, `DeployApp-${new Date().toISOString()}`, {
      sources: [Source.asset("../UI/dist/websocket-chat")],
      destinationBucket: siteBucket,
      distribution: distribution,
      distributionPaths: ['/'],
    });

    // Retrieving Cognito Userpool from existing resource - resolving circular stack dependency
    const userPool = UserPool.fromUserPoolId(this, "UserPool", props?.cognitoUserPoolId!);
    const appClient = userPool.addClient('websocket-frontend', {
      oAuth: {
        flows: {
          authorizationCodeGrant: false,
          implicitCodeGrant: true // return ID/Access tokens in returnURL
        },
        scopes: [OAuthScope.OPENID, OAuthScope.PROFILE, OAuthScope.EMAIL],
        callbackUrls: [`https://${distribution.distributionDomainName}/callback`, "http://localhost:4200/callback"],
        logoutUrls: [`https://${distribution.distributionDomainName}/login`, "http://localhost:4200/callback"],
      },

      idTokenValidity: Duration.minutes(720),
    });

    // Generate a cognito app client with a returnURL pointing to the Cloudfront distribution url
    const domain = userPool.addDomain('Domain', {
      cognitoDomain: {
        domainPrefix: props?.cognitoDomainPrefix!
      }
    });

    const cognitoSignInUrl = domain!.signInUrl(appClient!, {
      redirectUri: `https://${distribution.distributionDomainName}/callback`, // must be a URL configured under 'callbackUrls' with the client
    });

    const signinUrlParameter = new StringParameter(this, 'CognitoSigninURLParameter', {
      allowedPattern: '.*',
      description: 'Cognito Singin URL',
      parameterName: '/prod/cognito/signinurl',
      stringValue: cognitoSignInUrl,
      tier: ParameterTier.STANDARD,
    });

    const websocketUrlParameter = new StringParameter(this, 'WebsocketURLParameter', {
      allowedPattern: '.*',
      description: 'Websocket API URL',
      parameterName: '/prod/websocket/url',
      stringValue: `wss://${distribution.distributionDomainName}/wss/`,
      tier: ParameterTier.STANDARD,
    });

    const clientIdParameter = new StringParameter(this, 'CognitoClientIdParameter', {
      allowedPattern: '.*',
      description: 'Cognito client id',
      parameterName: '/prod/cognito/clientid',
      stringValue: appClient?.userPoolClientId!,
      tier: ParameterTier.STANDARD,
    });

    new CfnOutput(this, 'cognitoSigninURL', {
      value: cognitoSignInUrl,
      description: 'SignIn URL for Cognito Userpool',
      exportName: 'cognitoSigninURL',
    });
    new CfnOutput(this, 'DistributionId', { value: distribution.distributionId });
    new CfnOutput(this, 'DistributionURL', { value: distribution.distributionDomainName });
  }
};
